pkg/controller/datadependency/handler.go

```diff
@@ -0,0 +1,900 @@
+
+		// Attempt the update
+		_, updateErr := c.karmadaClient.WorkV1alpha2().ResourceBindings(rbToUpdate.Namespace).Update(context.TODO(), rbToUpdate, metav1.UpdateOptions{})
+		return updateErr
+	})
+}
+
+// injectPlacementAffinity injects placement affinity into a ResourceBinding based on DS locality
+// This function is only called for initial injection, so no need to handle previous exclusions
+func (c *DataDependencyController) injectPlacementAffinity(rb *workv1alpha2.ResourceBinding, dsClusterNames []string) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		latestRB, err := c.karmadaClient.WorkV1alpha2().ResourceBindings(rb.Namespace).Get(context.TODO(), rb.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		rbToUpdate := latestRB.DeepCopy()
+
+		// Add the annotation
+		if rbToUpdate.Annotations == nil {
+			rbToUpdate.Annotations = make(map[string]string)
+		}
+		rbToUpdate.Annotations[PlacementInjectedAnnotation] = "true"
+
+		// Ensure Placement exists
+		if rbToUpdate.Spec.Placement == nil {
+			rbToUpdate.Spec.Placement = &policyv1alpha1.Placement{}
+		}
+
+		// Set cluster affinity based on DS locality
+		if len(dsClusterNames) > 0 {
+			// Get all clusters first
+			allClusters, err := c.cLister.List(labels.Everything())
+			if err != nil {
+				return fmt.Errorf("failed to list all clusters: %v", err)
+			}
+
+			// Convert dsClusterNames to a set for efficient lookup
+			dsClusterSet := make(map[string]bool)
+			for _, clusterName := range dsClusterNames {
+				dsClusterSet[clusterName] = true
+			}
+
+			// Calculate excluded clusters (all clusters - DS clusters)
+			var excludedClusters []string
+			for _, cluster := range allClusters {
+				if !dsClusterSet[cluster.Name] {
+					excludedClusters = append(excludedClusters, cluster.Name)
+				}
+			}
+
+			// Merge with existing user-defined exclusions if any
+			var finalExcludedClusters []string
+			if rbToUpdate.Spec.Placement.ClusterAffinity != nil && len(rbToUpdate.Spec.Placement.ClusterAffinity.ExcludeClusters) > 0 {
+				// Keep existing user-defined exclusions
+				finalExcludedClusters = append(finalExcludedClusters, rbToUpdate.Spec.Placement.ClusterAffinity.ExcludeClusters...)
+			}
+
+			// Add our exclusions
+			finalExcludedClusters = append(finalExcludedClusters, excludedClusters...)
+
+			// Remove duplicates
+			excludedSet := make(map[string]bool)
+			var uniqueExcluded []string
+			for _, cluster := range finalExcludedClusters {
+				if !excludedSet[cluster] {
+					excludedSet[cluster] = true
+					uniqueExcluded = append(uniqueExcluded, cluster)
+				}
+			}
+
+			// Update ClusterAffinity
+			if rbToUpdate.Spec.Placement.ClusterAffinity == nil {
+				rbToUpdate.Spec.Placement.ClusterAffinity = &policyv1alpha1.ClusterAffinity{}
+			}
+			rbToUpdate.Spec.Placement.ClusterAffinity.ExcludeClusters = uniqueExcluded
```

Comment:
Why not just specify `ClusterNames` in `ClusterAffinity`? Do we should have to use `ExcludeClusters`?